### PR TITLE
[libc] Add Linux mman extension remap_file_pages.

### LINKED ISSUE #110122: [libc] add linux mman extension `remap_file_pages`
We need to create a `remap_file_pages` wrapper for linux targets, see https://www.man7.org/linux/man-pages/man2/remap_file_pages.2.html.

This requires:

- implement the entrypoint header similar to `libc/src/sys/mman/mmap.h`
   - modify `libc/src/sys/mman/CMakeLists.txt` accordingly
- implement the entrypoint similar to `libc/src/sys/mman/linux/mmap.cpp`
   - modify `libc/src/sys/mman/linux/CMakeLists.txt` accordingly
- add in the header spec
  - modify `functions` section in `libc/newhdrgen/yaml/sys/mman.yaml`
  - modify `SysMMan` section in `libc/spec/linux.td`
- add in test
  - similar to `libc/test/src/sys/mman/linux/mmap_test.cpp`
  - modify `libc/test/src/sys/mman/linux/CMakeLists.txt` accordingly

-----------------------------------------------------
Misc:
- How to build libc for dev: https://libc.llvm.org/full_host_build.html
- Code style and dev guide: https://libc.llvm.org/dev/index.html

If you use vscode, [sample settings](https://code.visualstudio.com/docs/getstarted/settings#_settings-json-file) are as the following (optional settings are useful in speeding up the compilation):
```
{
    "cmake.sourceDirectory": "${workspaceFolder}/runtimes",
    "cmake.configureSettings": {
        "CMAKE_EXPORT_COMPILE_COMMANDS": true,
        "LLVM_ENABLE_RUNTIMES": [
            "libc",
            "compiler-rt"
        ],
        "LLVM_LIBC_FULL_BUILD": true,
        "LLVM_USE_LINKER": "lld", // optional: requires installing lld
        "CMAKE_CXX_COMPILER_LAUNCHER": "sccache", // optional: requires installing sccache first
        "CMAKE_C_COMPILER_LAUNCHER": "sccache", // optional: requires installing sccache first
        "CMAKE_CXX_COMPILER": "/usr/bin/clang++",
        "CMAKE_C_COMPILER": "/usr/bin/clang",
        "LLVM_LIBC_INCLUDE_SCUDO": true,
        "COMPILER_RT_BUILD_SCUDO_STANDALONE_WITH_LLVM_LIBC": true,
        "COMPILER_RT_BUILD_GWP_ASAN": false,
        "COMPILER_RT_SCUDO_STANDALONE_BUILD_SHARED": false,
        "LIBC_USE_NEW_HEADER_GEN": true
    },
    "cmake.generator": "Ninja", // optional: requires installing Ninja
    "editor.formatOnSave": true,
    "files.insertFinalNewline": true,
    "clangd.arguments": [
        "--header-insertion=never"
    ]
}
```

### PR DESCRIPTION
**Pull request by AlyElashram**
Fixes https://github.com/llvm/llvm-project/issues/110122
- Create remap_file_pages.h/.cpp wrapper for the linux sys call.
- Add UnitTests for remap_file_pages
- Add function to libc/spec/linux.td
- Add Function spec to mman.yaml

### CODE REVIEW

**Comment by AlyElashram** [comment]
@SchrodingerZhu I'm not really certain of the testing macros used , I kind of "monkey see monkey do"-ed them from other test files, so please let me know if something is off.

File: libc/src/sys/mman/linux/remap_file_pages.cpp
Comment (michaelrj-google) [inline libc/src/sys/mman/linux/remap_file_pages.cpp:44]: nit: needs ending newline

File: libc/test/src/sys/mman/linux/remap_file_pages_test.cpp
Comment (michaelrj-google) [inline libc/test/src/sys/mman/linux/remap_file_pages_test.cpp:74]: nit: needs ending newline

File: libc/src/sys/mman/linux/remap_file_pages.cpp
Comment (michaelrj-google) [inline libc/src/sys/mman/linux/remap_file_pages.cpp:1]: nit: since this function isn't actually from POSIX, you can remove that from the title here.

File: libc/src/sys/mman/linux/remap_file_pages.cpp
Comment (michaelrj-google) [inline libc/src/sys/mman/linux/remap_file_pages.cpp:20]: nit: since this is a linux specific function you don't need this comment.

File: libc/src/sys/mman/linux/remap_file_pages.cpp
Comment (michaelrj-google) [inline libc/src/sys/mman/linux/remap_file_pages.cpp:25]: there's only one possible syscall here, so you can move the syscall into this ifdef like this (though you may need to reformat the code):
```suggestion
  int ret = LIBC_NAMESPACE::syscall_impl<int>(SYS_remap_file_pages, reinterpret_cast<long>(addr), size, prot, pgoff, flags);
```

File: libc/src/sys/mman/remap_file_pages.h
Comment (michaelrj-google) [inline libc/src/sys/mman/remap_file_pages.h:13]: we don't need `sys/mman.h` here, since this function doesn't take any types that it defines. Instead this should just be `stddef.h`
```suggestion
#include <stddef.h>
```

File: libc/src/sys/mman/linux/remap_file_pages.cpp
Comment (michaelrj-google) [inline libc/src/sys/mman/linux/remap_file_pages.cpp:9]: this should use `"quote marks"` instead of `<angle brackets>`
```suggestion
#include "src/sys/mman/remap_file_pages.h"
```

File: libc/newhdrgen/yaml/sys/mman.yaml
Comment (michaelrj-google) [inline libc/newhdrgen/yaml/sys/mman.yaml:103]: nit: in other places it's capitalized as `Linux`
```suggestion
      - Linux
```

**Review by michaelrj-google — COMMENTED** [review]
This looks like a good start, I left some comments for minor things to fix.

**Comment by AlyElashram** [comment]
I have one concern , I downloaded the logs for the previous build and didn't find the new remap_file_pages_test test run after searching for it. Shouldn't this be a concern? @michaelrj-google

**Comment by michaelrj-google** [comment]
Ah, you need to add this function to [`entrypoints.txt`](https://github.com/llvm/llvm-project/blob/main/libc/config/linux/x86_64/entrypoints.txt) which is the list of functions we provide for each target. If you want to see which functions are being skipped, add `-DLIBC_CMAKE_VERBOSE_LOGGING=ON` to your cmake command.

File: libc/src/sys/mman/linux/CMakeLists.txt
Comment (AlyElashram) [inline libc/src/sys/mman/linux/CMakeLists.txt:161]: this is going to cause an error most definitely.
```suggestion
    remap_file_pages.cpp
  HDRS
    ../remap_file_pages.h
```

**Comment by AlyElashram** [comment]
Converting this to a draft till I figure out why the tests are missing :eyes:

File: libc/test/src/sys/mman/linux/remap_file_pages_test.cpp
Comment (AlyElashram) [inline libc/test/src/sys/mman/linux/remap_file_pages_test.cpp:35]: I believe this makes much more sense when using remap_file_pages , instead of using MAP_PRIVATE and not having a file backed mapping. The use case of remap_file_pages should remap a file backed memory allocation if i understand correctly.

@michaelrj-google  @SchrodingerZhu  please correct me if I'm wrong here.

File: libc/test/src/sys/mman/linux/remap_file_pages_test.cpp
Comment (michaelrj-google) [inline libc/test/src/sys/mman/linux/remap_file_pages_test.cpp:28]: If you're opening a file in a test you should use a file in the same directory as the test, similar to what's done in [`access_test`](https://github.com/llvm/llvm-project/blob/main/libc/test/src/unistd/access_test.cpp#L26-L27)

also small nit: this should use `LIBC_NAMESPACE::open`.

File: libc/test/src/sys/mman/linux/remap_file_pages_test.cpp
Comment (michaelrj-google) [inline libc/test/src/sys/mman/linux/remap_file_pages_test.cpp:40]: you don't need to reset errno here since you asserted that it succeeded above.

File: libc/test/src/sys/mman/linux/remap_file_pages_test.cpp
Comment (michaelrj-google) [inline libc/test/src/sys/mman/linux/remap_file_pages_test.cpp:59]: same file path stuff as above, though make sure to use a different name since tests may be run in parallel.

**Comment by AlyElashram** [comment]
@michaelrj-google @SchrodingerZhu Pinging on this as I think it's good to go. Awaiting approval to rebase 👀

**Comment by SchrodingerZhu** [comment]
Hi, I’m traveling right now but I will take a look after my flight today. Michael and others are bit busy currently as they are preparing for the upcoming devconf. Thanks for your patience and apologies for the delay.
Yifan

 Sent from Proton Mail for iOS 
On Sat, Oct 12, 2024 at 15:26, Aly ElAshram ***@***.***&gt; wrote:   @michaelrj-google @SchrodingerZhu Pinging on this as I think it's good to go. Awaiting approval to rebase 👀 —

File: libc/test/src/sys/mman/linux/CMakeLists.txt
Comment (SchrodingerZhu) [inline libc/test/src/sys/mman/linux/CMakeLists.txt:138]:
```suggestion
    libc.include.sys_mman
    libc.include.sys_stat
```

**Review by SchrodingerZhu — CHANGES_REQUESTED** [review]

File: libc/test/src/sys/mman/linux/CMakeLists.txt
Comment (AlyElashram) [reply libc/test/src/sys/mman/linux/CMakeLists.txt:138]: @SchrodingerZhu  Adding it in the next commit , but can I know why we are adding sys_stat?

**Comment by AlyElashram** [comment]
Should I rebase and push --force-with-lease after getting 2 approvals ? or should I do it now @SchrodingerZhu ?

**Review by SchrodingerZhu — APPROVED** [review]

**Comment by SchrodingerZhu** [comment]
> Should I rebase and push --force-with-lease after getting 2 approvals ? or should I do it now @SchrodingerZhu ?

I guess no more operation on your side is needed for now. let's wait to see if the CI is happy.

**Pull request merged**
